### ISSUE TITLE
Balance of Bionics Slots

### ISSUE THREAD

**Issue by illi-kun** [CleverRaven/Cataclysm-DDA#16920]
As we discussed this with @Coolthulhu in #16915, it will be good to try to balance bionics slots before enabling the restrictions.

Current list of occupied slots for all bionics is presented below. These numbers were defined without detailed comparison, so maybe you have some propositions how to make it more reasonable / better, I'm all ears.

The same table is also available in csv format: 
[cbms_slots.csv.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/287777/cbms_slots.csv.zip)

Edit by @Leland: put list in collapsable form

<details>
 <summary>List Of Occupied Slots</summary>

| CBM Name | Bodypart | Slot |
| --- | --- | --- |
| Acidic Discharge | ARM_L | 1 |
| Acidic Discharge | ARM_R | 1 |
| Acidic Discharge | LEG_L | 2 |
| Acidic Discharge | LEG_R | 2 |
| Acidic Discharge | TORSO | 3 |
| Active Defense System | ARM_L | 1 |
| Active Defense System | ARM_R | 1 |
| Active Defense System | HEAD | 1 |
| Active Defense System | LEG_L | 2 |
| Active Defense System | LEG_R | 2 |
| Active Defense System | TORSO | 10 |
| Adrenaline Pump | TORSO | 6 |
| Advanced Microreactor System | TORSO | 55 |
| Aero-Evaporator | HAND_R | 1 |
| Aero-Evaporator | TORSO | 5 |
| Air Filtration System | MOUTH | 2 |
| Air Filtration System | TORSO | 4 |
| Alarm System | HEAD | 1 |
| Alarm System | TORSO | 2 |
| Alloy Plating - Arms | ARM_L | 4 |
| Alloy Plating - Arms | ARM_R | 4 |
| Alloy Plating - Head | HEAD | 5 |
| Alloy Plating - Legs | LEG_L | 6 |
| Alloy Plating - Legs | LEG_R | 6 |
| Alloy Plating - Torso | TORSO | 10 |
| Anti-glare Compensators | EYES | 1 |
| Artificial Night Generator | TORSO | 16 |
| Battery System | TORSO | 25 |
| Bionic Claws | HAND_L | 5 |
| Bionic Claws | HAND_R | 5 |
| Bionic Nostril | HEAD | 2 |
| Bionic Nostril | MOUTH | 1 |
| Bionic Short Circuit | ARM_L | 2 |
| Bionic Short Circuit | ARM_R | 2 |
| Bionic Short Circuit | LEG_L | 3 |
| Bionic Short Circuit | LEG_R | 3 |
| Bionic Short Circuit | TORSO | 1 |
| Bionic Visual Impairment | EYES | 1 |
| Bionic-Induced Deformity | HEAD | 4 |
| Bionic-Induced Deformity | MOUTH | 1 |
| Blood Analysis | TORSO | 5 |
| Blood Filter | TORSO | 10 |
| Cerebral Booster | HEAD | 7 |
| Chain Lightning | ARM_L | 4 |
| Chain Lightning | ARM_R | 4 |
| Chain Lightning | TORSO | 16 |
| Cloaking System | ARM_L | 3 |
| Cloaking System | ARM_R | 3 |
| Cloaking System | FOOT_L | 1 |
| Cloaking System | FOOT_R | 1 |
| Cloaking System | HAND_L | 1 |
| Cloaking System | HAND_R | 1 |
| Cloaking System | HEAD | 2 |
| Cloaking System | LEG_L | 4 |
| Cloaking System | LEG_R | 4 |
| Cloaking System | TORSO | 15 |
| Close Quarters Battle | HEAD | 4 |
| Cranial Flashlight | EYES | 1 |
| Diamond Cornea | EYES | 1 |
| Dielectric Capacitance System | ARM_L | 2 |
| Dielectric Capacitance System | ARM_R | 2 |
| Dielectric Capacitance System | FOOT_L | 1 |
| Dielectric Capacitance System | FOOT_R | 1 |
| Dielectric Capacitance System | HAND_L | 1 |
| Dielectric Capacitance System | HAND_R | 1 |
| Dielectric Capacitance System | HEAD | 2 |
| Dielectric Capacitance System | LEG_L | 3 |
| Dielectric Capacitance System | LEG_R | 3 |
| Dielectric Capacitance System | TORSO | 8 |
| Directional EMP | ARM_R | 1 |
| Directional EMP | HAND_R | 2 |
| Electrical Discharge | TORSO | 6 |
| Electrical Drain | TORSO | 2 |
| Electromagnetic Unit | HAND_R | 3 |
| Electroshock Unit | ARM_L | 3 |
| Electroshock Unit | ARM_R | 3 |
| Electroshock Unit | HAND_L | 1 |
| Electroshock Unit | HAND_R | 1 |
| Electroshock Unit | TORSO | 8 |
| Endocrine Enervator | HEAD | 2 |
| Enhanced Hearing | HEAD | 3 |
| Enhanced Memory Banks | HEAD | 3 |
| Ethanol Burner | TORSO | 26 |
| Expanded Digestive System | TORSO | 20 |
| Facial Distortion | HEAD | 3 |
| Facial Distortion | MOUTH | 1 |
| Finger-Mounted Laser | HAND_R | 1 |
| Fingerhack | HAND_R | 2 |
| Fingerpick | HAND_R | 2 |
| Fingertip Razors | HAND_L | 2 |
| Fingertip Razors | HAND_R | 2 |
| Flashbang Generator | ARM_L | 2 |
| Flashbang Generator | ARM_R | 2 |
| Flashbang Generator | LEG_L | 3 |
| Flashbang Generator | LEG_R | 3 |
| Flashbang Generator | TORSO | 20 |
| Fusion Blaster Arm | ARM_L | 20 |
| Fusion Blaster Arm | HAND_L | 5 |
| Heat Drain | ARM_L | 3 |
| Heat Drain | ARM_R | 3 |
| Heat Drain | FOOT_L | 2 |
| Heat Drain | FOOT_R | 2 |
| Heat Drain | HAND_L | 1 |
| Heat Drain | HAND_R | 1 |
| Heat Drain | HEAD | 3 |
| Heat Drain | LEG_L | 5 |
| Heat Drain | LEG_R | 5 |
| Heat Drain | TORSO | 12 |
| Hydraulic Muscles | ARM_L | 8 |
| Hydraulic Muscles | ARM_R | 8 |
| Hydraulic Muscles | LEG_L | 14 |
| Hydraulic Muscles | LEG_R | 14 |
| Hydraulic Muscles | TORSO | 4 |
| Implanted Night Vision | EYES | 1 |
| Infrared Vision | EYES | 1 |
| Integrated Dosimeter | ARM_L | 1 |
| Integrated Dosimeter | ARM_R | 1 |
| Integrated Dosimeter | HEAD | 1 |
| Integrated Dosimeter | LEG_L | 1 |
| Integrated Dosimeter | LEG_R | 1 |
| Integrated Dosimeter | TORSO | 2 |
| Integrated Toolset | HAND_L | 3 |
| Integrated Toolset | HAND_R | 3 |
| Internal Chronometer | HEAD | 1 |
| Internal Climate Control | ARM_L | 2 |
| Internal Climate Control | ARM_R | 2 |
| Internal Climate Control | FOOT_L | 1 |
| Internal Climate Control | FOOT_R | 1 |
| Internal Climate Control | HAND_L | 1 |
| Internal Climate Control | HAND_R | 1 |
| Internal Climate Control | HEAD | 2 |
| Internal Climate Control | LEG_L | 3 |
| Internal Climate Control | LEG_R | 3 |
| Internal Climate Control | TORSO | 8 |
| Internal Furnace | TORSO | 40 |
| Internal Storage | TORSO | 32 |
| Internal Unified Power System | TORSO | 5 |
| Itchy Metal Thing | HAND_L | 1 |
| Itchy Metal Thing | HAND_R | 1 |
| Itchy Metal Thing | HEAD | 2 |
| Itchy Metal Thing | TORSO | 3 |
| Joint Torsion Ratchet | ARM_L | 8 |
| Joint Torsion Ratchet | ARM_R | 8 |
| Joint Torsion Ratchet | LEG_L | 12 |
| Joint Torsion Ratchet | LEG_R | 12 |
| Leaky Bionic | TORSO | 5 |
| Leukocyte Breeder System | TORSO | 10 |
| Membrane Oxygenator | HEAD | 2 |
| Membrane Oxygenator | MOUTH | 2 |
| Membrane Oxygenator | TORSO | 4 |
| Metabolic Interchange | TORSO | 25 |
| Microreactor System | TORSO | 55 |
| Mini-Flamethrower | HAND_L | 1 |
| Mini-Flamethrower | HAND_R | 1 |
| Monomolecular Blade | ARM_R | 5 |
| Monomolecular Blade | HAND_R | 2 |
| Motor Control Overstimulator | HEAD | 3 |
| Muscle Augmentation | ARM_L | 4 |
| Muscle Augmentation | ARM_R | 4 |
| Muscle Augmentation | LEG_L | 8 |
| Muscle Augmentation | LEG_R | 8 |
| Muscle Augmentation | TORSO | 6 |
| Nictating Membrane | EYES | 1 |
| Noisemaker | ARM_L | 1 |
| Noisemaker | ARM_R | 1 |
| Noisemaker | LEG_L | 1 |
| Noisemaker | LEG_R | 1 |
| Noisemaker | TORSO | 4 |
| Offensive Defense System | ARM_L | 1 |
| Offensive Defense System | ARM_R | 1 |
| Offensive Defense System | HEAD | 1 |
| Offensive Defense System | LEG_L | 2 |
| Offensive Defense System | LEG_R | 2 |
| Offensive Defense System | TORSO | 10 |
| Olfactory Mask | ARM_L | 1 |
| Olfactory Mask | ARM_R | 1 |
| Olfactory Mask | HEAD | 1 |
| Olfactory Mask | LEG_L | 1 |
| Olfactory Mask | LEG_R | 1 |
| Olfactory Mask | TORSO | 3 |
| Optical Dampers | EYES | 1 |
| Plutonium Filter | TORSO | 10 |
| Plutonium Purger | TORSO | 10 |
| Power Armor Interface | HEAD | 1 |
| Power Armor Interface | TORSO | 4 |
| Power Armor Interface Mk. II | HEAD | 2 |
| Power Armor Interface Mk. II | TORSO | 3 |
| Power Overload | ARM_L | 1 |
| Power Overload | ARM_R | 1 |
| Power Overload | LEG_L | 2 |
| Power Overload | LEG_R | 2 |
| Power Overload | TORSO | 3 |
| Probability Travel | ARM_L | 2 |
| Probability Travel | ARM_R | 2 |
| Probability Travel | FOOT_L | 1 |
| Probability Travel | FOOT_R | 1 |
| Probability Travel | LEG_L | 3 |
| Probability Travel | LEG_R | 3 |
| Probability Travel | TORSO | 20 |
| Protective Lenses | EYES | 1 |
| Radiation Scrubber System | ARM_L | 1 |
| Radiation Scrubber System | ARM_R | 1 |
| Radiation Scrubber System | LEG_L | 2 |
| Radiation Scrubber System | LEG_R | 2 |
| Radiation Scrubber System | TORSO | 5 |
| Railgun | ARM_L | 5 |
| Railgun | ARM_R | 5 |
| Railgun | HAND_L | 1 |
| Railgun | HAND_R | 1 |
| Recycler Unit | TORSO | 15 |
| Remote Controller | HEAD | 2 |
| Repair Nanobots | TORSO | 10 |
| Scent Vision | HEAD | 3 |
| Self-Locking Thumbs | HAND_L | 1 |
| Self-Locking Thumbs | HAND_R | 1 |
| Sensory Dulling | HEAD | 2 |
| Shockwave Generator | TORSO | 20 |
| Solar Panels | TORSO | 20 |
| Sonic Resonator | ARM_L | 2 |
| Sonic Resonator | ARM_R | 2 |
| Sonic Resonator | HEAD | 1 |
| Sonic Resonator | LEG_L | 4 |
| Sonic Resonator | LEG_R | 4 |
| Sonic Resonator | TORSO | 15 |
| Sound Dampeners | HEAD | 3 |
| Squeaky Ankles | LEG_L | 3 |
| Squeaky Ankles | LEG_R | 3 |
| Subdermal Carbon Filament | ARM_L | 2 |
| Subdermal Carbon Filament | ARM_R | 2 |
| Subdermal Carbon Filament | FOOT_L | 1 |
| Subdermal Carbon Filament | FOOT_R | 1 |
| Subdermal Carbon Filament | HAND_L | 1 |
| Subdermal Carbon Filament | HAND_R | 1 |
| Subdermal Carbon Filament | HEAD | 2 |
| Subdermal Carbon Filament | LEG_L | 3 |
| Subdermal Carbon Filament | LEG_R | 3 |
| Subdermal Carbon Filament | TORSO | 5 |
| Synaptic Accelerator | ARM_L | 3 |
| Synaptic Accelerator | ARM_R | 3 |
| Synaptic Accelerator | FOOT_L | 2 |
| Synaptic Accelerator | FOOT_R | 2 |
| Synaptic Accelerator | HAND_L | 1 |
| Synaptic Accelerator | HAND_R | 1 |
| Synaptic Accelerator | LEG_L | 6 |
| Synaptic Accelerator | LEG_R | 6 |
| Synaptic Accelerator | TORSO | 12 |
| Targeting System | ARM_L | 3 |
| Targeting System | ARM_R | 3 |
| Targeting System | EYES | 1 |
| Targeting System | HAND_L | 1 |
| Targeting System | HAND_R | 1 |
| Teleportation Unit | ARM_L | 3 |
| Teleportation Unit | ARM_R | 3 |
| Teleportation Unit | LEG_L | 4 |
| Teleportation Unit | LEG_R | 4 |
| Teleportation Unit | TORSO | 16 |
| Telescopic Eyes | EYES | 2 |
| Terranian Sonar | FOOT_L | 3 |
| Terranian Sonar | FOOT_R | 3 |
| Thermal Dissipation | ARM_L | 2 |
| Thermal Dissipation | ARM_R | 2 |
| Thermal Dissipation | FOOT_L | 1 |
| Thermal Dissipation | FOOT_R | 1 |
| Thermal Dissipation | HAND_L | 1 |
| Thermal Dissipation | HAND_R | 1 |
| Thermal Dissipation | HEAD | 2 |
| Thermal Dissipation | LEG_L | 3 |
| Thermal Dissipation | LEG_R | 3 |
| Thermal Dissipation | TORSO | 10 |
| Time Dilation | ARM_L | 3 |
| Time Dilation | ARM_R | 3 |
| Time Dilation | FOOT_L | 1 |
| Time Dilation | FOOT_R | 1 |
| Time Dilation | HAND_L | 1 |
| Time Dilation | HAND_R | 1 |
| Time Dilation | HEAD | 3 |
| Time Dilation | LEG_L | 4 |
| Time Dilation | LEG_R | 4 |
| Time Dilation | TORSO | 12 |
| Uncanny Dodge | ARM_L | 3 |
| Uncanny Dodge | ARM_R | 3 |
| Uncanny Dodge | FOOT_L | 1 |
| Uncanny Dodge | FOOT_R | 1 |
| Uncanny Dodge | LEG_L | 5 |
| Uncanny Dodge | LEG_R | 5 |
| Uncanny Dodge | TORSO | 12 |
| Visual Disruptor | EYES | 1 |
| Voice Remodulator | MOUTH | 1 |
| Voice Remodulator | TORSO | 2 |
| Water Extraction Unit | ARM_R | 1 |
| Water Extraction Unit | HAND_R | 2 |
| Weather Reader | HEAD | 1 |
| Weather Reader | TORSO | 1 |
| Wire-induced Stiffness | ARM_L | 1 |
| Wire-induced Stiffness | ARM_R | 1 |
| Wire-induced Stiffness | FOOT_L | 1 |
| Wire-induced Stiffness | FOOT_R | 1 |
| Wire-induced Stiffness | LEG_L | 1 |
| Wire-induced Stiffness | LEG_R | 1 |
| Wire-induced Stiffness | TORSO | 2 |
| Wired Reflexes | ARM_L | 4 |
| Wired Reflexes | ARM_R | 4 |
| Wired Reflexes | LEG_L | 8 |
| Wired Reflexes | LEG_R | 8 |
| Wired Reflexes | TORSO | 6 |
</summary>

**Comment by sanyaskillpro** [comment]
Microreactor is way too bulky. You spend valuable consumables to fuel it, also it takes 2 other CBMs to upgrade it, i think that's enough downsides. It should be on par with internal furnace.

Same with internal storage. 32 torso slots just for 8 storage?

**Comment by Coolthulhu** [comment]
Faulty bionics don't need to cost slots, they are bad enough on their own

Those are good bionics that need to cost a lot of slots:
* Speed booster
* Stat boosters
* Targeting system
* Time stop - instant win in most combat situations due to power storage bionics not costing slots
* Repair nanobots - this one can pretty much guarantee immunity to hp damage
* Sensory dulling - this one guarantees near immunity to pain - near because it causes death if used to negate 200+ pain
* Blood filter - alone it isn't anything special, but with dulling it removes that 200+ pain risk thing
* Internal UPS
* Dielectric
* Cloaking
* Monoblade
* Hydraulic
* EMP - costs no time to use
* Mini-flamethrower - also no time to use
* Leukocyte breeder - breaks the health system and guarantees perfect health

Out of the above, only hydraulic really has anywhere near the slot cost it should have.

Energy generation bionics, by usefulness:
* Metabolic is the most convenient, unless player mutates carnivore or very fast metabolism. It uses cheap resources and can work in any condition. Should have the highest or the second highest cost
* Furnace works fast, uses very cheap resources and can be used in almost every condition (except very intense combat). Roughly tied with metabolic.
* Ratchet works in most conditions and costs nothing to use, but is rather slow
* Solar panels only work during the day, but they're free to use. Perfect for regeneration, not as good for combat or lab diving
* Ethanol uses cheap resources and works fast, but alcohol causes debuffs and can result in addiction that has to be cold turkey'd out
* Battery system can be used at any time and works very fast, but consumes limited resources and thus doesn't help with crafting
* Reactor consumes expensive resources and is unreliable. Until power storage start costing slots, it is the weakest energy generation bionic

Those bionics are rather weak and should be cheaper:
* Heat drain (pure crap, should only cost hand slots)
* Internal storage
* Reactor + extra components it comes with. In fact, the extra components should be free and just count as parts of the reactor
* Armor bionics
* Dosimeter. It duplicates functionality of a 4 volume item

**Comment by illi-kun** [comment]
> Faulty bionics don't need to cost slots, they are bad enough on their own

I like how faulty bionics balance Faulty Cyborg profession and I'm strongly disagree with making _any_ CBMs to have 0-slots but I agree to make it significantly cheaper.

**Comment by illi-kun** [comment]
@Coolthulhu concerning the test, it's good input and I'll apply it.

@sanyaskillpro your comment seems reasonable to me. Such big numbers for reactor, internal storage and some other bionics were set just to represent that previously we had hardcoded condition that these CBMs can't be installed together. It's possible to revise it now.

**Comment by Coolthulhu** [comment]
> I'm strongly disagree with making any CBMs to have 0-slots

That's certainly a wrong approach in case of bundled bionics, like dampeners that go with hearing. Hidden costs are inexcusable bad design here - inexcusable because it can be trivially fixed.

Also, it will result in faulty installations making even less sense than before - trying to install a hand slot bionic resulting in head slot being wasted.

Finally, it prevents the slot system from being consistent - slot limits will not be guaranteed, preventing balancing around that.

> I like how faulty bionics balance Faulty Cyborg profession

They balance it by making it weaker. Having it at the same time suck at everything and be unable to actually use the bionics it starts with will not do anything good to the balance of this profession.

**Comment by pisskop** [comment]
Failed bionics still occupy the space.  They still conume presumed connections and limit what else can be placed.

Maybe failed cyborgs need less failed parts instead of so many leaky, squeaky, and freaky bionics?

**Comment by baraba3** [comment]
Faulty bionics should occupy space but it should be possible to remove them with first aid kit.

Internal chronometer shoudl occupy 0 slots as its practicly useless as watches are abundant.Or it should recive some additional funcion like alarm strong enough to wake heavy sleeper or something along those lines.

**Comment by Voskhod-DDA** [comment]
I'd say that faulty bionics should still take up slots, but for those number of slots to be reduced.
I'd also like to suggest a work-around to the problem of failed installations - How about each faulty bionic corresponds to a certain CBM? Such as Wired Reflexes/Joint Torsion Ratchet/ turning into Wire Induced Stiffness on failure, rather than just a random faulty CBM? Finger-razortips would on failure become self-locking thumbs, failure to install any of the eye-based CBMs would result in Bionic Visual Impairment, and so on and so on.

**Comment by pisskop** [comment]
isnt the recipe for batteries a potato, can, and scrap metal?

all easily obtained in endgame

**Comment by Voskhod-DDA** [comment]
Doesn't help that you can just eat said potato to get ~200 bionic power from the nutrition

**Comment by pisskop** [comment]
Presumably it would be a small attatchment.  Its not weak though.  lets not assume that inconvenience means not-useful

**Comment by kevingranade** [comment]
Please, do not balance number of slots with how "useful" you think the
bionic is, if it's a bad bionic, it can just be a bad bionic, and if it's
good, it makes more sense to give it meaningful drawbacks than to
arbitrarily make it smaller.
The number of slots should be based on how large and/or invasive it is.

**Comment by illi-kun** [comment]
@kevingranade that was my initial approach. In this case we need to balance cbms by rarity of its appearance, so really good bionics should spawn rarer than common one.

Also, judgment about size of each cbm is quite subjective because all we have for making such conclusion is bionic description.

**Comment by TempestMaiden** [comment]
There's a bunch of bionics that only take up slots in one arm or hand; in every case it will be from the limb on the right. Is it intended to let the player choose which side to install them on, eventually?

Also, I think the solar panel CBM takes up too much torso space - compare it to Subdermal Carbon Filament or Torso Armor, both of which implicitly have total surface coverage. How deep are those panels going?

**Comment by illi-kun** [comment]
> Is it intended to let the player choose which side to install them on, eventually?

Yes, it will be added later.

**Comment by pingpong2011** [comment]
I'm pretty buggered by joint torsion ratchet. Works 100%, no cost except moving, and has no limitations except it's slow I guess? With a UPS CBM I can use multiple things free of charge - the electric jackhammer would probably the most egregious, costing only 10 power to begin with. A UPS flashlight costs two volume and provides more than twice the brightness of the cranial CBM for example, though the brightness is a drawback to using it in labs. MP3 player becomes free. If it weren't for a car being overpowered, you could use it with any tools. Even the integrated toolset would basically be free.

Unless arm CBMs were buffed, like lightning had a near 100% chance to stun an enemy, I don't see the point in breaking the flow of combat or increasing stuff carried to use many arm CBMs, especially if they cost turns you could be doing normal actions. Did they ever "fix" the fusion arm to not bounce off of a lot of things?

I don't know how much coal or other high-fuel stuff is valued for internal furnace, but eating two-by-fours and clothing has it's limitations. Eating is a zero-cost action though if I recall correctly, so it's doable in combat (metabolic interchange takes time, and critically, high density foods or cannibalism).

Any way if you're using a UPS-enhanced item, they take time to recharge, so speed of you recharging isn't as much of a limitation.

**Comment by TempestMaiden** [comment]
Most "arm cbms" are useful, always-on passives, or Hydraulic Muscles.

And Metabolic Interchange gives 25 power for every two units of nutrition. Unless you're trying to use it to power the timestop cbm, that's not a significant cost. A single slice of pizza gives you double the power Joint Torsion Ratchet will over the course of an hour - and faster, too. Metabolic Interchange gives an average of five power per turn.

**Comment by pingpong2011** [comment]
200 power / 46m42s for my character(117 base, 100 cost). Does 100 time count as 5 seconds? Presuming that since my character takes 6 seconds to move, comes out to around 5 power a minute, moving back and forth. It doesn't sound like much but unless you're constantly in battle, it provides far more than enough energy by just exploring for anything you'll do, including reading at night by headlamp.

**Comment by TempestMaiden** [comment]
You ever try to run Enhanced Memory Banks off of a torsion ratchet?

A slow trickle of power, just enough to run a few minor things constantly or for the rare short burst, is the exact and only thing the Joint Torsion Ratchet does. If you want to use anything actually power-hungry - MP3 player and Flashlight definitely don't count - the JTR doesn't cut it.

**Comment by kevingranade** [comment]
Several things:
Sounds like Torsion ratchet is working exactly as intended. It's also cheap
and crappy, yet totally reliable compared to the alternatives.
Metabolic interchange is definitely producing too much power, AND eating
should not be a 0-move action.

**Comment by mugling** [comment]
Where are we up to on this? It would be a shame not to enable @illi-kun's code?

**Comment by illi-kun** [comment]
What if amount of slots for each body part will be changed to 100, so number of slots of each bionic will means % of total available space?

**Comment by mugling** [comment]
That could make comparing some bionics a little wierd

**Comment by illi-kun** [comment]
Yeah, eyes CBMs with size 20/100 is not the same as torso CBM with size 20/100 but we can hide the total slots completely and show occupied size as 20%.

**Comment by illi-kun** [comment]
Note: balancing of armor bionics depends on #16739. It's possible to keep current slot size but tweak their resistance (or vice versa).

**Comment by Coolthulhu** [comment]
Having slots using same "volume units" (like now) would be useful for power storage, which is the biggest problem with unlimited CBMs.

If all slots used percentage, we'd either need a set of scaling values or a different system for power storage.

**Comment by illi-kun** [comment]
> Having slots using same "volume units" (like now) would be useful for power storage

Curent slots are not exact "volume units" since eyes total slots (4) equal mouth total slots (4) and almost equal to hand slots (5), and physical volume of these body parts is not so close.

If way of "volume units" is the preferable one, I want to set the total number of slots of entire body (except eyes and mount, since these are parts of head) as follows:
```Σslots(all bodyparts) - slots(eyes) - slots(mouth) = N * 62.5L,``` where 62.5L is a volume of medium (human) body size (as per current code), and ```N``` here is some reasonable factor (about 4 or 5).

In this case, cost of ```internal storage``` can be easily defined as its internal volume multiplied by factor ```N```. Also, the general benefit is the possibility to split our one-size CBMs installation packs to some groups (small CBMs, middle CBMs, large CBMs, so ```volume of CBM pack >= number of occupied slots / N```).

In top of that, I'm thinking about the way to define number of bionics we need to allow to install in one body: it's something in the midle between "you can install any CBMs you find" and "you need to carefully choose the only one bionic you will install in your body". What if sum of total number of required slots of all bionics will be ```X``` times bigger than total number of slots to use, so for example (_NB: the following numbers are random!_), if we have
- 10 different eyes CBMs;
- 4 total available slots in eyes;

then assuming ```X = 3```, we have average slots cost of each eyes CBM is equal to ```4 * 3 / 10 = 1.2 slot units```; of course, slots are integer and not all CBMs have equal size and are equally useful, so some of them cost more and some less but sum of occupied slots for all CBMs (which set in JSON and allowed to be installed in eyes) is 12 (so only third part of all CBMs can be installed to one human body). 

Any comments on this?

**Comment by keyspace** [comment]
Balancing by "usefullness" seems subjective, and overly complicated due to need of comparing sets.

Percentage/volume allow more down-to-earth reasoning, as demonstrated by @illi-kun for `bio_storage` above.

Could be done for some bionics with minor edits to description, e.g. `bio_blaster` to take 100% of hand and 50% of arm.

Does leave contention for some, like `bio_teleport`, but that's a smaller set than "all".

-----

Following probably tangent.

IMO balancing by percentage is tautological to balancing by volume - the latter would be used for reference eventually. It would also bring up the question of "why do CBMs in inventory take up the same space?.."

If volume-based slots are a go, could one be used to calculate the other?

Currently:

* not-yet-installed CBM item `volume` is set to 10 for all non-broken ones - in `data/json/items/bionics.json`; broken ones don't have a corresponding item;
* `occupied_bodyparts` of installed ones (broken and not) is in `data/json/bionics.json`.

**Comment by keyspace** [comment]
Ran [this script](https://gist.github.com/keyspace/8c3e3b4c06625944bb07fb104c9b54b6) on current `bionics.json`, it produced [this table](https://gist.github.com/keyspace/43a3459f7759971c253f1ab9512716e0) - might be useful for review.

The table doesn't fit into GH's `.contaner` HTML element for me, so increased its `width` to `1500` manually. Zooming out might also work.

**Comment by DangerNoodle** [comment]
The first obvious step would be ensuring that all existing professions do not start over any bionics limits. If so then something needs to be balanced, whether that be the bionic size or the profession.

Faulty Cyborg is actually rather little-affected by the size of their faulty bionics. As expected, Broken Cyborg is most affected by it, but they still fall within the limits, with their legs being the only slot maxed out. I am still testing each profession to ensure that their bionics are within limits.

I would however suggest that the size of faulty bionics be reduced, not necessarily eliminated. They already carry overt penalties on their own, so any space occupied is a mere nod to realism, not a balancing factor. Balancing factors that are a drop in the bucket compared to existing balancing features are unwarranted. See initial implementation of filthy clothing.

**Comment by Coolthulhu** [comment]
> If way of "volume units" is the preferable one, I want to set the total number of slots of entire body (except eyes and mount, since these are parts of head) as follows:
Σslots(all bodyparts) - slots(eyes) - slots(mouth) = N * 62.5L, where 62.5L is a volume of medium (human) body size (as per current code), and N here is some reasonable factor (about 4 or 5).

That sounds like big numbers on everything. Smaller numbers are preferable for readability and easier math.
I don't think we need huge granularity here - any bionic that takes less than 1/10 of any body part it occupies is getting into needless detail.

Tying it to average human body size sounds like a good start, but it could also lead to situation where someone wants to link it to current body size sum, which is something we want to avoid. Plus, it doesn't really reflect the situation all that well, as every body part has different amount of free space.

Rescaling the numbers won't fix them, though.
The only real reason to keep the units consistent is power storage. They're the only bionics that are likely to be installed in different body parts. If power storage could only occupy torso, the percentage units could be better than volume units.

We need to figure out how much power storage should a cyborg have.
Let's start with 1000 maximum - on the safe side, but still low enough to make power not free.
Any cases where it isn't enough?

Then, we need some character builds that should be doable and some that should not. Balancing things without examples will quickly turn into "but on paper it sounds good/bad".

**Comment by kevingranade** [comment]
I'd actually say start with the 'this should be fine'/'this is too much'
examples, the rest of the system would just enforce the distinctions
present in the examples.

**Comment by kevingranade** [comment]
Stale issue is stale.
